document how the custom annotations work (#1564)
aws/aws-advanced-jdbc-wrapper: examples/SpringHibernateBalancedReaderOneDataSourceExample/CUSTOM_ANNOTATIONS.md

# Custom Annotations Documentation

## Overview

The SpringHibernateBalancedReaderOneDataSourceExample demonstrates a **single datasource approach** for load-balanced read-write and read-only connections using the AWS Advanced JDBC Driver's built-in Read/Write Splitting Plugin. Unlike the TwoDataSourceExample, this approach does **not** use custom annotations for datasource routing.

## Key Difference from TwoDataSourceExample

| Aspect | OneDataSourceExample | TwoDataSourceExample |
|--------|---------------------|----------------------|
| **Datasources** | Single datasource | Two separate datasources (writer + reader) |

> | **Routing Mechanism** | AWS JDBC Driver's Read/Write Splitting Plugin | Custom `@WithLoadBalancedReaderDataSource` annotation |

Reviewer: ```suggestion
| **Routing Mechanism** | AWS Advanced JDBC Wrapper's Read/Write Splitting Plugin | Custom `@WithLoadBalancedReaderDataSource` annotation |
```

| **Configuration** | Connection string with `wrapperProfileName=F0` | Spring's `AbstractRoutingDataSource` + AOP |
| **Custom Annotations** | None (uses standard Spring annotations) | Custom annotation with AspectJ interception |
| **Complexity** | Simpler, driver-managed | More complex, application-managed |

## Standard Annotations Used

### @Transactional(readOnly = true)

The key to routing in this example is Spring's standard `@Transactional` annotation with the `readOnly` parameter:

```java

> @Retryable(value = { ShouldRetryTransactionException.class }, maxAttempts = 3)

Reviewer 2: Probably need to replace `ShouldRetryTransactionException` with `FailoverSQLException`

@Transactional(propagation = Propagation.REQUIRES_NEW, readOnly = true)
public int getNumOfBooksTransactional() {

>     // AWS JDBC Driver automatically routes this to a reader instance

Reviewer 2: ```suggestion
    // AWS Advanced JDBC Wrapper automatically routes this to a reader instance
```

### Custom Exception Translation

The `Config` class includes a custom `HibernateExceptionTranslator` to handle failover exceptions:

```java
@Bean
public HibernateExceptionTranslator hibernateExceptionTranslator(){
  return new HibernateExceptionTranslator() {
    @Override
    public DataAccessException translateExceptionIfPossible(RuntimeException ex) {
      // Convert JDBC connection exceptions to retryable exceptions
      if (ex.getCause() != null && ex.getCause() instanceof JDBCConnectionException) {
        return new ShouldRetryTransactionException(ex);
      }
      return super.translateExceptionIfPossible(ex);
    }
  };
}
```

**Purpose:**
- Converts `JDBCConnectionException` (including failover exceptions) to `ShouldRetryTransactionException`
- Enables `@Retryable` to automatically retry failed transactions
- Provides seamless failover handling

## Complete Flow

### Read Operation Flow

1. **Method Call**: Application calls a method with `@Transactional(readOnly = true)`
2. **Transaction Start**: Spring starts a read-only transaction
3. **Connection Request**: Hibernate requests a connection from the datasource

> 4. **Driver Detection**: AWS JDBC Driver detects the read-only transaction flag

Reviewer 2: ```suggestion
4. **Driver Detection**: AWS Advanced JDBC Wrapper detects the read-only transaction flag
```

## Best Practices

1. **Always Use readOnly Flag**: Explicitly set `readOnly = true` for read operations
   ```java
   @Transactional(readOnly = true)
   ```

2. **Use Configuration Profiles**: Leverage predefined profiles (F0, D, E, F) for optimal settings
   ```
   wrapperProfileName=F0
   ```

3. **Configure Reader Selection**: Choose appropriate reader selection strategy
   ```
   readerHostSelectorStrategy=roundRobin  // or random, leastConnections
   ```

4. **Enable Retry Logic**: Always use `@Retryable` for failover resilience
   ```java
   @Retryable(value = { ShouldRetryTransactionException.class }, maxAttempts = 3)
   ```

5. **Custom Exception Translation**: Implement `HibernateExceptionTranslator` to convert failover exceptions

## Example: Complete Method Annotations

### Read-Only Method
```java
@Retryable(value = { ShouldRetryTransactionException.class }, maxAttempts = 3)
@Transactional(propagation = Propagation.REQUIRES_NEW, readOnly = true)
public int getNumOfBooksTransactional() {
    // Automatically routed to reader instance
    return this.repository.findAll().stream().mapToInt(Book::getQuantityAvailable).sum();
}
```

### Read-Write Method
```java

> @Retryable(value = { ShouldRetryTransactionException.class, TransactionSystemException.class}, maxAttempts = 3)

Reviewer: ```suggestion
@Retryable(value = {ShouldRetryTransactionException.class, TransactionSystemException.class}, maxAttempts = 3)
```
